### PR TITLE
stitch relations after deleting

### LINKED ISSUE #13774: Stale hasMember relation left behind on Group after deleting User
## Expected Behavior

Relations created automatically during earlier processing are trimmed if the cause for them to be created no longer exists.

## Actual Behavior

Stale `hasMember` relations continue to exist on a Group after the User it refers to (that did claim `memberOf` that group) is deleted, either via UI or API deletion.

## Steps to Reproduce

1. Add the YAML below to somewhere your Backstage instance can ingest from (I used GitHub Discovery directed at a repo with YAML files)
2. Wait for processing & observe the relations on the `test-group` are created despite the omission of a `members` value.
3. Comment out `billy.someone`
4. Wait for processing and check `billy.someone` gets annotated as Orphaned.
5. Delete the user via the UI "Deregister" button or via an API call to `DELETE /entities/by-uid/<uid>`.
6. Observe the `test-group` relations remain unchanged.

```yaml
apiVersion: backstage.io/v1alpha1
kind: Group
metadata:
  name: test-group
spec:
  type: test
  children: []
---
apiVersion: backstage.io/v1alpha1
kind: User
metadata:
  name: billy.someone
spec:
  memberOf: [test-group]
---
apiVersion: backstage.io/v1alpha1
kind: User
metadata:
  name: simple.simon
spec:
  memberOf: [test-group]
```

Note: The same issue occurs even if you provide a correct or incorrect `members` value, but we run our test setup without currently.

## Context

We planed to use the username part of a user's email as their User entity's `metadata.name`, but this issue might expose us to a GDPR violation as that name would remain in the DB even after we delete any leaving employees.

## Your Environment

Backstage is running from containers in AWS EKS (kubernetes), and the backing database is Postgres 14 from AWS RDS.

Output of `yarn backstage-cli info` as gathered on my laptop as the command doesn't seem to work inside the container, built from your example Dockerfile so running Debian 11.  But the node/yarn versions matched with the below.

Note I first observed this in v1.6.0-next2, but only started playing with Backstage recently and & couldn't say if this issue was present in earlier versions or not.

```text
node: v16.17.0
yarn: 1.22.19
cli:  0.19.0 (installed)
backstage:  1.6.0

Dependencies:
  @backstage/app-defaults                          1.0.6
  @backstage/backend-common                        0.15.1
  @backstage/backend-plugin-api                    0.1.2
  @backstage/backend-tasks                         0.3.5
  @backstage/catalog-client                        1.1.0
  @backstage/catalog-model                         1.1.1
  @backstage/cli-common                            0.1.10
  @backstage/cli                                   0.19.0
  @backstage/config-loader                         1.1.4
  @backstage/config                                1.0.2
  @backstage/core-app-api                          1.1.0
  @backstage/core-components                       0.11.1
  @backstage/core-plugin-api                       1.0.6
  @backstage/errors                                1.1.1
  @backstage/integration-react                     1.1.4
  @backstage/integration                           1.3.1
  @backstage/plugin-api-docs                       0.8.9
  @backstage/plugin-app-backend                    0.3.36
  @backstage/plugin-auth-backend                   0.16.0
  @backstage/plugin-auth-node                      0.2.5
  @backstage/plugin-catalog-backend-module-aws     0.1.9
  @backstage/plugin-catalog-backend-module-github  0.1.7
  @backstage/plugin-catalog-backend                1.4.0
  @backstage/plugin-catalog-common                 1.0.6
  @backstage/plugin-catalog-graph                  0.2.21
  @backstage/plugin-catalog-import                 0.8.12
  @backstage/plugin-catalog-node                   1.1.0
  @backstage/plugin-catalog-react                  1.1.4
  @backstage/plugin-catalog                        1.5.1
  @backstage/plugin-circleci                       0.3.9
  @backstage/plugin-github-actions                 0.5.9
  @backstage/plugin-home                           0.4.25
  @backstage/plugin-kubernetes-backend             0.7.2
  @backstage/plugin-kubernetes-common              0.4.2
  @backstage/plugin-kubernetes                     0.7.2
  @backstage/plugin-org                            0.5.9
  @backstage/plugin-permission-common              0.6.4
  @backstage/plugin-permission-node                0.6.5
  @backstage/plugin-permission-react               0.4.5
  @backstage/plugin-proxy-backend                  0.2.30
  @backstage/plugin-scaffolder-backend             1.6.0
  @backstage/plugin-scaffolder-common              1.2.0
  @backstage/plugin-scaffolder                     1.6.0
  @backstage/plugin-search-backend-module-pg       0.4.0
  @backstage/plugin-search-backend-node            1.0.2
  @backstage/plugin-search-backend                 1.0.2
  @backstage/plugin-search-common                  1.0.1
  @backstage/plugin-search-react                   1.1.0
  @backstage/plugin-search                         1.0.2
  @backstage/plugin-sentry                         0.4.2
  @backstage/plugin-stack-overflow                 0.1.5
  @backstage/plugin-techdocs-backend               1.3.0
  @backstage/plugin-techdocs-module-addons-contrib 1.0.4
  @backstage/plugin-techdocs-node                  1.4.0
  @backstage/plugin-techdocs-react                 1.0.4
  @backstage/plugin-techdocs                       1.3.2
  @backstage/plugin-user-settings                  0.4.8
  @backstage/release-manifests                     0.0.6
  @backstage/test-utils                            1.2.0
  @backstage/theme                                 0.2.16
  @backstage/types                                 1.0.0
  @backstage/version-bridge                        1.0.1
Done in 1.08s.
```

### PR DESCRIPTION
**Pull request by freben**
Fixes #13774

### CODE REVIEW

**Review by jhaals — APPROVED** [review]
Nice work freben! 🌟 

Can we do something about relations that might be dangling prior to this fix?

File: plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts [hunk 1]
```diff
@@ -256,9 +262,35 @@ export class DefaultEntitiesCatalog implements EntitiesCatalog {
...
+          .andWhere('refresh_state.entity_id', '!=', uid)
+          .select({ ref: 'relations.source_entity_ref' }),
+      );
+
+    // Perform the actual deletion
     await this.database<DbRefreshStateRow>('refresh_state')
       .where('entity_id', uid)
       .delete();
+
+    await this.stitcher.stitch(new Set(relationPeers.map(p => p.ref)));
```
Comment (Rugvip) [inline plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts:293]: We need to do this for `replaceUnprocessedEntities` as well right?

**Comment by freben** [comment]
@jhaals Hmm not really ... what people could do is to mass-set the right hash field to just some junk string which would trigger stitching on the next processing round. That'd get the entire database into a properly stitched state. The tricky thing is, we don't know easily which ones are just dangling because they are being legitimately generated that way (e.g. if your spec points to an owner that doesn't exist - which is "legal" and will keep being generated like that even if we deleted it), and which ones are dangling because of a partial deletion.

File: plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts (hunk 1 above)
Comment (freben) [reply plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts:293]: Maybe - will have to look into that separately in that case, gotta jump on another thing first

Comment (Rugvip) [reply plugins/catalog-backend/src/service/DefaultEntitiesCatalog.ts:293]: A TODO? :grin:

**Pull request merged**